src/worker.rs

```diff
@@ -0,0 +1,320 @@
+    /// otherwise the parent supervisor will block and eventually fail with a
+    /// timeout.
+    ///
+    /// ### Report a start error with the StartNotifier
+    ///
+    /// If for some reason, a worker is not able to start correctly (e.g. DB
+    /// connection fails, network is kaput, etc.), the worker may call the
+    /// `StartNotifier#failed` function with the impending error as a parameter.
+    /// This will cause the whole supervision system start procedure to abort
+    /// and fail fast.
+    ///
+    pub fn new_with_start<F, O>(name: &str, mut routine0: F) -> Self
+    where
+        F: FnMut(Context, StartNotifier) -> O + 'static,
+        O: Future<Output = anyhow::Result<()>> + FutureExt + Send + Sized + 'static,
+    {
+        let routine1 = move |ctx: Context, on_start: StartNotifier| routine0(ctx, on_start).boxed();
+        Spec {
+            name: name.to_owned(),
+            shutdown: Shutdown::Indefinitely,
+            restart: Restart::Permanent,
+            routine: Box::new(routine1),
+        }
+    }
+
+    /// restart is a builder method that allows to modify the Restart settings
+    /// for the spawned worker
+    pub fn restart(mut self, restart: Restart) -> Self {
+        self.restart = restart;
+        self
+    }
+
+    /// shutdown is a builder method that allows to modify the Restart settings
+    /// for the spawned worker
+    pub fn shutdown(mut self, shutdown: Shutdown) -> Self {
+        self.shutdown = shutdown;
+        self
+    }
+
+    /// start spawns a new routine that executes the start function of this
+    /// Spec.
+    ///
+    /// ### Blocking on start function
+    ///
+    /// This function will block and wait until a notification from the
+    /// start function (executing on a newly spawned routine) is received; this
+    /// notification indicates if the start was successful or not.
+    ///
+    /// This blocking is necessary to ensure that supervised worker trees get
+    /// spawned in the specified/expected order.
+    pub async fn start(
+        mut self,
+        parent_ctx: &Context,
+        parent_name: &str,
+    ) -> anyhow::Result<Worker> {
+        let runtime_name = format!("{}/{}", parent_name, self.name);
+        let created_at = Utc::now();
+
+        // ON_START setup
+        let (started_tx, started_rx) = oneshot::channel::<Option<anyhow::Error>>();
+        let start_notifier = StartNotifier::from_oneshot(started_tx);
+
+        // CANCEL setup
+        let (ctx, termination_handle) = Context::with_cancel(parent_ctx);
+        let (kill_handle, kill_registration) = AbortHandle::new_pair();
+        let future = (*self.routine)(ctx, start_notifier);
+        let routine = Abortable::new(future, kill_registration);
+
+        // SPAWN -- actually spawn concurrent worker future
+        let join_handle = task::spawn(routine);
+
+        // ON_START blocking -- block before moving to the next start
+        let err = started_rx.await;
```

Comment:
Is it really right to wait on this oneshot indefinitely though? The documentation of `new_with_start` mentions that the a worker that never reports starting will be killed after a timeout. I don't see that in the code yet, so I assume that is still to-be-implemented?

Reply:
The intent is to make a timeout here, I completely missed that on the first step though, I think is worth doing that work in this PR. Not to difficult to add a timeout call here and also add a test for that.